targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F429xI/TARGET_NUCLEO_F429ZI/PinNames.h

```diff
@@ -38,6 +38,22 @@
 extern "C" {
 #endif
 
+#define STDIO_UART               UART_3
+
+    // STDIO for console print
+#ifdef MBED_CONF_TARGET_STDIO_UART_TX
+#  define STDIO_UART_TX MBED_CONF_TARGET_STDIO_UART_TX
```

Comment:
Because I'm using macros to select how to console should be initialized. If all data required for explicit pin-map is defined, then we use explicit pin-map to init console:
https://github.com/ARMmbed/mbed-os/blob/48ad43a2fc75dddf6ced2db760ada7fa7fd4320b/platform/source/mbed_retarget.cpp#L282

But now we could use `constexpr` utility functions to do the job and remove these `FUNC` macros.

Reply:
No #define STDIO_UART_TX_FUNC in case of MBED_CONF_TARGET_STDIO_UART_TX ?

Reply:
`MBED_CONF_TARGET_STDIO_UART_XX` comes from the config file, so in such case, we would have to extend the config file. I think that these extra macros can be removed now and we can use `constexpr` utility functions to find mappings. I'm checking this now.

Reply:
> If all data required for explicit pin-map is defined, then we use explicit pin-map to init console:

Gotcha. The inability to test for enums (plus people using enums) drives me nuts in HAL-type code.

I don't particularly fancy the maintenance burden of maintaining the pinmap tables AND those separate defines. I think I'd kind of like to just pressure people to do the full constexpr thing - "if you have the constexpr stuff, retarget will use it, so your image will get smaller".

I did a parallel define thing for the us_ticker optimisations, and that's also ugly, but there is no "plan B" alternative yet...

If those defines do stay, I'd want to, if possible, have a cross-check that they're correct. Have a Greentea test that checks the run-time lookup for the pins matches any declared defines. That was done for the us_ticker optimisation - defines vs us_ticker_info structure.

Reply:
I tried to use `constexpr` pin-map functions to init the console and I found a strange issue.

I got two versions of code:
```
    const serial_pinmap_t console_pinmap = get_uart_pinmap(STDIO_UART_TX, STDIO_UART_RX);
    static DirectSerial console(console_pinmap, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
```

```
    static DirectSerial console(get_uart_pinmap(STDIO_UART_TX, STDIO_UART_RX), MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
```

I can see the following changes in the used flash when the second form is used:
- `ARM: +392 bytes!`
- `GCC_ARM: +16 bytes`
- `IAR -16 bytes`


Details for `ARM` compiler below:
```
| Module                          |       .text |   .data |       .bss |
|---------------------------------|-------------|---------|------------|
| [lib]\c_w.l                     |   11175(+0) |  16(+0) |    348(+0) |
| [lib]\fz_wm.l                   |      34(+0) |   0(+0) |      0(+0) |
| [lib]\m_wm.l                    |      48(+0) |   0(+0) |      0(+0) |
| anon$$obj.o                     |      32(+0) |   0(+0) | 197888(+0) |
| drivers\source                  |     192(+0) |   0(+0) |      0(+0) |
| features\netsocket              |     143(+0) |   0(+0) |      0(+0) |
| hal\mbed_critical_section_api.o |     154(+0) |   0(+0) |      2(+0) |
| hal\mbed_gpio.o                 |      96(+0) |   0(+0) |      0(+0) |
| hal\mbed_ticker_api.o           |     978(+0) |   0(+0) |      0(+0) |
| hal\mbed_us_ticker_api.o        |     114(+0) |   4(+0) |     65(+0) |
| main.o                          |      70(+0) |   0(+0) |      0(+0) |
| platform\source                 |  5683(-392) |  64(+0) |    249(+0) |
| rtos\source                     |    8990(+0) | 168(+0) |   6626(+0) |
| targets\TARGET_Freescale        |   16581(+0) |  12(+0) |    340(+0) |
| Subtotals                       | 44290(-392) | 264(+0) | 205518(+0) |
Total Static RAM memory (data + bss): 205782(+0) bytes
Total Flash memory (text + data): 44554(-392) bytes
```

Any ideas?

Reply:
I covered this in passing in a comment I made an hour ago. The second form prompts the compiler to generate code to put the (precomputed) `spi_pinmap_t` on the stack and passes a stack pointer to the function. The first form just has the precomputed pinmap in ROM and points to it.

So the first form can generate smaller code. Not sure how IAR makes it bigger though.

It may be a technical requirement of the C++ standard that it goes onto the stack - distinct objects may have to have distinct addresses, so it can't just make all calls with that pinmap point to the same thing.

(String literals can share addresses, as can C99 constant compound literals IIRC, but C++ constant temporaries can't. I think).

Reply:
Oh, but you'd want to make sure that separate `spi_pinmap_t` was `static` too, or it might put it on the stack anyway, if that's inside a function.

Reply:
```
    static const serial_pinmap_t console_pinmap = get_uart_pinmap(STDIO_UART_TX, STDIO_UART_RX);
    static DirectSerial console(console_pinmap, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
```

Adding `static` gives:
- `ARM: -22 B`
- `GCC_ARM: +0 B`
- `IAR: -16 B`

Reply:
I found a problem while working on the serial console initialization using `constexpr` pin-map function. The solution works fine, but I lost backward compatibility. 
At the moment console is initialized using explicit pin-map mechanism (without constexpr) if all required defines are available: `STDIO_UART_TX, STDIO_UART_TX_FUNC, STDIO_UART_RX, STDIO_UART_RX_FUNC, STDIO_UART`. Otherwise, we use the old version which uses pin-map tables. 

Now when I'm using `contsexpr` function to find mapping I need to have `peripheralPinMaps.h` files ready for all targets. 
Maybe a good solution would be to add some configuration macro for targets which support explicit pin-maps in `targets.json`. Something like: `device_has: "EXPLICIT_PINMAP"` ?
This could be removed when explicit pin-map support is added to all targets, but for now, I could add support in parts: golden, silver, bronze, other targets.

Reply:
I'd put a indicator macro in PinNames.h that you have the extended HAL functionality (that was new macros in device.h. That's similar to what I did in #10609. Don't really need to drag the JSON into it, as it's an extension to something you can already find the main header for.